### PR TITLE
Feature/proximity

### PR DESCRIPTION
**Pull request by BNS-MarkUlrich**

### CODE REVIEW

File: Assets/Scripts/Proximity/ProximityAgent.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            if (range <= 0)
+                range = _defaultRange;
+
+            _nearbyAgents = this.GetAgentsInRange(range, agentTypesToCheck);
+            BroadCastNearbyAgents(_nearbyAgents.ToArray());
+
+#region Testing
+            // foreach (ProximityAgent agent in _nearbyAgents)
+            //     Debug.DrawLine(transform.position, agent.transform.position, Color.black);
+#endregion
```
Comment (swzwij) [inline Assets/Scripts/Proximity/ProximityAgent.cs]: Are the testing regions still needed?

File: Assets/Scripts/Proximity/ProximityManager.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+using FishNet.Object;
+using Unity.VisualScripting;
+using UnityEngine;
+
+namespace Examen.Proximity
+{
+    public static class ProximityManager
+    {
+        private static readonly Dictionary<ProximityAgent, AgentTypes> _proximityAgents = new();
+        private static readonly Dictionary<AgentTypes, List<ProximityAgent>> _proximityAgentsByType = new(); // Test later
```
Comment (swzwij) [inline Assets/Scripts/Proximity/ProximityManager.cs]: "Test later" ?

File: Assets/Scripts/Proximity/ProximityAgent.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            if (!IsServer)
+                return;
+
+            if (!_checkProximity)
+                return;
+
+            GetProximityData(_defaultRange, _agentTypesToCheck);
+        }
+
+        public HashSet<ProximityAgent> RequestProximityData(float range = 0, params AgentTypes[] agentTypesToCheck)
```
Comment (swzwij) [inline Assets/Scripts/Proximity/ProximityAgent.cs:44]: Missing summary :)

File: Assets/Scripts/Proximity/ProximityManager.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+        public static void Subscribe(this ProximityAgent agent, AgentTypes agentType)
```
Comment (swzwij) [inline Assets/Scripts/Proximity/ProximityManager.cs:17]: Missing summary :)

File: Assets/Scripts/Proximity/ProximityManager.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            if (_proximityAgents.TryGetValue(agent, out _))
+                return;
+
+            _proximityAgents.Add(agent, agentType);
+        }
+
+        public static void Unsubscribe(this ProximityAgent agent)
```
Comment (swzwij) [inline Assets/Scripts/Proximity/ProximityManager.cs:29]: Missing summary :)

File: Assets/Scripts/Proximity/ProximityManager.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            if (_proximityAgents.TryGetValue(agent, out _))
+                _proximityAgents.Remove(agent);
+        }
+
+        public static HashSet<ProximityAgent> GetAgentsOfType(AgentTypes agentTypes)
```
Comment (swzwij) [inline Assets/Scripts/Proximity/ProximityManager.cs:40]: Missing summary :)

File: Assets/Scripts/Proximity/ProximityManager.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            foreach (var entity in _proximityAgents)
+            {
+                if (entity.Value == agentTypes)
+                    entities.Add(entity.Key);
+            }
+
+            return entities;
+        }
+
+        public static HashSet<ProximityAgent> GetAgentsInRange(this ProximityAgent sourceAgent, float range, params AgentTypes[] agentTypes)
```
Comment (swzwij) [inline Assets/Scripts/Proximity/ProximityManager.cs:60]: Missing summary :)

File: Assets/Scripts/Proximity/TestingProximity.cs [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+using System.Linq;
+using FishNet.Object;
+using UnityEngine;
+
+namespace Examen.Proximity
+{
+    public class TestingProximity : NetworkBehaviour
+    {
+        //[SerializeField] private AgentTypes[] _agentTypesToCheck;
+        //[SerializeField] private float _range = 15f;
```
Comment (swzwij) [inline Assets/Scripts/Proximity/TestingProximity.cs]: Remove comments

File: Assets/Scripts/Proximity/TestingProximity.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        {
+            _proximityAgent = GetComponent<ProximityAgent>();
+            StartCoroutine(CheckProximity(1f));
+        }
+
+        private IEnumerator CheckProximity(float interval)
+        {
+            while (true)
+            {
+                //_nearbyAgents = _proximityAgent.RequestProximityData(_range, _agentTypesToCheck).ToArray();
```
Comment (swzwij) [inline Assets/Scripts/Proximity/TestingProximity.cs]: Remove comments

**Review by swzwij — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Proximity/ProximityManager.cs (hunk 2 above)
Comment (BNS-MarkUlrich) [reply Assets/Scripts/Proximity/ProximityManager.cs]: Shush

File: Assets/Scripts/Proximity/ProximityManager.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        /// <summary>
+        /// Retrieves all ProximityAgents of the specified AgentTypes.
+        /// </summary>
+        /// <param name="agentTypes">The AgentTypes to filter the ProximityAgents by.</param>
+        /// <returns>A HashSet of ProximityAgents of the specified AgentTypes.</returns>
+        public static HashSet<ProximityAgent> GetAgentsOfType(AgentTypes agentTypes)
+        {
+            HashSet<ProximityAgent> entities = new();
+
+            foreach (var entity in _proximityAgents)
```
Comment (swzwij) [inline Assets/Scripts/Proximity/ProximityManager.cs]: ew var

**Review by swzwij — CHANGES_REQUESTED** [review]

**Review by swzwij — APPROVED** [review]

**Pull request merged**